libraries/chain/protocol/restriction.cpp

```diff
@@ -0,0 +1,551 @@
+   }
+
+   template<typename T>
+   void validate_by_member_type( const char* name, const extension<T>* t )
+   {
+      // extract the underlying type
+      validate_by_member_type( name, (const T*)nullptr );
+   }
+
+   template<typename T>
+   void validate_by_member_type( const char* name, const vector<T>* t )
+   {
+      validate_list_like_member<T>( name );
+   }
+
+   template<typename T>
+   void validate_by_member_type( const char* name, const set<T>* t )
+   {
+      validate_list_like_member<T>( name );
+   }
+
+   template<typename T>
+   void validate_by_member_type( const char* name, const flat_set<T>* t )
+   {
+      validate_list_like_member<T>( name );
+   }
+
+   template<typename... T>
+   void validate_by_member_type( const char* name, const flat_map<T...>* t )
+   {
+      FC_THROW( "Restriction on '${name}' is not supported due to its type",
+                ("name", name) );
+   }
+
+   template<typename... T>
+   void validate_by_member_type( const char* name, const static_variant<T...>* t )
+   {
+      FC_THROW( "Restriction on '${name}' is not supported due to its type",
+                ("name", name) );
+   }
+
+   void validate_by_member_type( const char* name, const fba_accumulator_id_type* t )
+   {
+      FC_THROW( "Restriction on '${name}' is not supported due to its type",
+                ("name", name) );
+   }
+
+};
+
+struct member_validation_visitor
+{
+   typedef void result_type;
+
+   const restriction& _restriction;
+
+   member_validation_visitor( const restriction& r ) : _restriction(r) {}
+
+   template<typename Member, class Class, Member (Class::*member)>
+   void operator()( const char* name )const
+   {
+      restriction_validation_helper helper( _restriction );
+      helper.validate_by_member_type( name, (const Member*)nullptr );
+   }
+};
+
+struct restriction_validation_visitor
+{
+   typedef void result_type;
+   const restriction& _restriction;
+
+   restriction_validation_visitor( const restriction& r ) : _restriction(r) {}
+
+   template<typename OpType>
+   result_type operator()( const OpType& ) // Note: the parameter is a reference of *nullptr, we should not use it
+   {
+      FC_ASSERT( _restriction.member < fc::reflector<OpType>::total_member_count,
+                 "member number ${m} is too large",
+                 ("m",_restriction.member) );
+
+      // other member modifiers should have been checked outside, so only check mmod_none here
+      if( _restriction.member_modifier.value == restriction::mmod_none )
+      {
+         member_validation_visitor vtor( _restriction );
+         fc::reflector<OpType>::visit_local_member( vtor, _restriction.member.value );
+      }
+   }
+
+};
+
+#define GRAPHENE_OP_IDX_CASE_VISIT(r, data, I, elem) \
+   case I : \
+      operation::visit< I >( vtor ); \
+      break;
+
+void validate_restriction_details( const restriction& r, unsigned_int op_type )
+{
+   restriction_validation_visitor vtor( r );
+   switch( op_type.value )
```

Comment:
Need to refactor due to changes on `static_variant`: https://github.com/bitshares/bitshares-fc/pull/79